lib/Sema/TypeCheckDistributed.cpp

```diff
@@ -761,19 +761,34 @@ void TypeChecker::checkDistributedActor(SourceFile *SF, NominalTypeDecl *nominal
           auto candidates = nominal->lookupDirect(func->getName());
           if (candidates.size() > 1) {
             auto firstDecl = dyn_cast<AbstractFunctionDecl>(candidates.back());
-            for (auto decl : candidates) {
-              if (decl == firstDecl) {
-                decl->diagnose(
-                    diag::distributed_func_cannot_overload_on_async_only,
-                    decl->getName());
-              } else {
-                decl->diagnose(
-                    diag::distributed_func_other_ambiguous_overload_here,
-                    decl->getName());
+            for (auto candidate: candidates) {
+              if (auto candidateFunc = dyn_cast<AbstractFunctionDecl>(candidate)) {
+                assert(candidateFunc->getParameters()->size() ==
+                       firstDecl->getParameters()->size());
+                bool allSame = true;
+                for (size_t i = 0; i < candidateFunc->getParameters()->size(); ++i) {
+                  auto lhs = firstDecl->getParameters()->get(i);
+                  auto rhs = candidateFunc->getParameters()->get(i);
+                  if (!lhs->getInterfaceType()->isEqual(rhs->getInterfaceType())) {
```

Comment:
This makes me wonder whether instead we should augment re-declaration checker to discount `async` and `throws` on `distributed func`, I think that would give us the behavior we want here instead of trying to equate stuff in ad-hoc way i.e. it would diagnose
```
distributed func test<T>(a: T) {}
distrbiuted func test<U>(a: U) async {}
```

With a custom message instead of the default "invalid redeclaration of".

Reply:
Yeah maybe that's easier hmmm... I'll see if I can find where that is done.

Reply:
Very nice, that simplified things a lot!